server/manager/src/main/java/org/apache/accumulo/manager/Manager.java

```diff
@@ -280,38 +280,42 @@ Fate<Manager> fate() {
       /* UNLOAD_ROOT_TABLET */      {O, O, O, X, X, X, X},
       /* STOP */                    {O, O, O, O, O, X, X}};
   //@formatter:on
-  synchronized void setManagerState(ManagerState newState) {
-    if (state.equals(newState)) {
+  synchronized void setManagerState(final ManagerState newState) {
+    if (state == newState) {
       return;
     }
     if (!transitionOK[state.ordinal()][newState.ordinal()]) {
       log.error("Programmer error: manager should not transition from {} to {}", state, newState);
     }
-    ManagerState oldState = state;
+    final ManagerState oldState = state;
     state = newState;
     nextEvent.event("State changed from %s to %s", oldState, newState);
-    if (newState == ManagerState.STOP) {
-      // Give the server a little time before shutdown so the client
-      // thread requesting the stop can return
-      ScheduledFuture<?> future = getContext().getScheduledExecutor().scheduleWithFixedDelay(() -> {
-        // This frees the main thread and will cause the manager to exit
-        clientService.stop();
-        Manager.this.nextEvent.event("stopped event loop");
-      }, 100L, 1000L, MILLISECONDS);
-      ThreadPools.watchNonCriticalScheduledTask(future);
-    }
-
-    if (oldState != newState && (newState == ManagerState.HAVE_LOCK)) {
-      upgradeCoordinator.upgradeZookeeper(getContext(), nextEvent);
-    }
-
-    if (oldState != newState && (newState == ManagerState.NORMAL)) {
-      if (fateRef.get() != null) {
-        throw new IllegalStateException("Access to Fate should not have been"
-            + " initialized prior to the Manager finishing upgrades. Please save"
-            + " all logs and file a bug.");
-      }
-      upgradeMetadataFuture = upgradeCoordinator.upgradeMetadata(getContext(), nextEvent);
+    switch (newState) {
+      case STOP:
+        // Give the server a little time before shutdown so the client
+        // thread requesting the stop can return
+        final var future = getContext().getScheduledExecutor().scheduleWithFixedDelay(() -> {
+          // This frees the main thread and will cause the manager to exit
+          clientService.stop();
+          Manager.this.nextEvent.event("stopped event loop");
+        }, 100L, 1000L, MILLISECONDS);
+        ThreadPools.watchNonCriticalScheduledTask(future);
+        break;
+      case HAVE_LOCK:
+        upgradeCoordinator.upgradeZookeeper(getContext(), nextEvent);
+        break;
+      case NORMAL:
+        // fate isn't started until after upgrades are done, so non-null fateRef implies isUpgrading
+        // should return false
+        if (fateRef.get() != null && isUpgrading()) {
```

Comment:
The manager starts off assuming an upgrade is not happening, thinking this assumption could be improved by changing the following variable creation to start off as true.  Then the manager would assume an upgrade is happening until it gets positive evidence otherwise.  Not sure if that change breaks anything, it would make this validation a bit stronger.

```java
private final AtomicBoolean managerUpgrading = new AtomicBoolean(true);
```

Alternatively the instance variable `managerUpgrading` could be dropped and the method ` isUpgrading()` changed to have the following impl.   This would have the same effect where the manager assumes an upgrade is in progress from startup until it determines its done or not needed.

```java
  @Override
  public boolean isUpgrading() {
    return upgradeCoordinator.getStatus() != UpgradeCoordinator.UpgradeStatus.COMPLETE;
  }
```

Reply:
I ended up going the second route with this. It made things much nicer to reason about, and I could use isUpgrading for more strict checks elsewhere.